torch/_overrides.py

```diff
@@ -0,0 +1,242 @@
+
+
+_wrapped_func_source = textwrap.dedent("""
+    @functools.wraps(implementation)
+    def {name}(*args, **kwargs):
+        relevant_args = dispatcher(*args, **kwargs)
+        return implement_torch_function(
+            implementation, {name}, relevant_args, args, kwargs)
+    """)
+
+
+def torch_function_dispatch(dispatcher, module=None, verify=True,
+                            docs_from_dispatcher=False):
+    """Decorator for adding dispatch with the __torch_function__ protocol.
+
+    See for example ``torch/functional.py`` for usage examples.
+
+    Parameters
+    ----------
+    dispatcher : callable
+        Function that when called like ``dispatcher(*args, **kwargs)`` with
+        arguments from the NumPy function call returns an iterable of
+        array-like arguments to check for ``__torch_function__``.
+    module : str, optional
+        ``__module__`` attribute to set on new function, e.g.,
+        ``module='torch'``.  By default, module is copied from the decorated
+        function.
+    verify : bool, optional
+        If True, verify the that the signature of the dispatcher and decorated
+        function signatures match exactly: all required and optional arguments
+        should appear in order with the same names, but the default values for
+        all optional arguments should be ``None``. Only disable verification
+        if the dispatcher's signature needs to deviate for some particular
+        reason, e.g., because the function has a signature like
+        ``func(*args, **kwargs)``.
+    docs_from_dispatcher : bool, optional
+        If True, copy docs from the dispatcher function onto the dispatched
+        function, rather than from the implementation. This is useful for
+        functions defined in C, which otherwise don't have docstrings.
+
+    Returns
+    -------
+    dispatcher : callable
+        Function suitable for decorating the implementation of a NumPy
+        function.
+
+    Notes
+    -----
+    The dispatcher should normally return a tuple containing all input
+    arguments that may have a ``__torch_function__`` attribute.
+
+    In some cases where that's not easily possible, e.g. ``torch.cat``, it is
+    also valid (if a little slower) to make the dispatcher function a generator
+    (i.e. use ``yield`` to return arguments one by one).
+
+    """
+    def decorator(implementation):
+        if verify:
+            verify_matching_signatures(implementation, dispatcher)
+
+        if docs_from_dispatcher:
+            add_docstring(implementation, dispatcher.__doc__)
+
+        # Equivalently, we could define this function directly instead of using
+        # exec. This version has the advantage of giving the helper function a
+        # more interpretable name. Otherwise, the original function does not
+        # show up at all in many cases, e.g., if it's written in C++ or if the
+        # dispatcher gets an invalid keyword argument.
```

Comment:
Hmm. Does overriding `__repr__` on the decorator type work?

Reply:
There's a test that `Tensor.__repr__` works as expected, and the `DiagonalTensor` test has a custom `__repr__`. And overriding:
```
In [1]: import torch                                                               

In [2]: torch.unique.__repr__()                                                    
Out[2]: '<function unique at 0x7fb935427378>'

In [3]: torch.unique.__repr__ = lambda : 'aahhhh'                                  

In [4]: torch.unique.__repr__()                                                    
Out[4]: 'aahhhh'

In [5]: torch.unique is torch.unique._implementation  # checking the decorator was active                                                                     
Out[5]: False
```